presto-native-execution/presto_cpp/main/PeriodicServiceInventoryManager.cpp

```diff
@@ -0,0 +1,138 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "presto_cpp/main/PeriodicServiceInventoryManager.h"
+#include <folly/futures/Retrying.h>
+#include <velox/common/memory/Memory.h>
+
+namespace facebook::presto {
+PeriodicServiceInventoryManager::PeriodicServiceInventoryManager(
+    std::string address,
+    int port,
+    std::shared_ptr<CoordinatorDiscoverer> coordinatorDiscoverer,
+    std::string clientCertAndKeyPath,
+    std::string ciphers,
+    std::string id,
+    uint64_t frequencyMs)
+    : address_(std::move(address)),
+      port_(port),
+      coordinatorDiscoverer_(std::move(coordinatorDiscoverer)),
+      clientCertAndKeyPath_(std::move(clientCertAndKeyPath)),
+      ciphers_(std::move(ciphers)),
+      id_(std::move(id)),
+      frequencyMs_(frequencyMs),
+      pool_(velox::memory::addDefaultLeafMemoryPool(id_)),
+      eventBaseThread_(false /*autostart*/) {}
+
+void PeriodicServiceInventoryManager::start() {
+  eventBaseThread_.start(id_);
+  stopped_ = false;
+  auto* eventBase = eventBaseThread_.getEventBase();
+  eventBase->runOnDestruction([this] { client_.reset(); });
+  eventBase->schedule([this]() { return sendRequest(); });
+}
+
+void PeriodicServiceInventoryManager::stop() {
+  stopped_ = true;
+  eventBaseThread_.stop();
+}
+
+void PeriodicServiceInventoryManager::sendRequest() {
+  // stop() calls EventBase's destructor which executed all pending callbacks;
+  // make sure not to do anything if that's the case
+  if (stopped_) {
+    return;
+  }
+
+  try {
+    folly::SocketAddress newAddress = coordinatorDiscoverer_->getAddress();
+    // Update service address after `updateServiceTimes()` attempts.
+    if (updateServiceTimes() > 0 && (attempts_++) % updateServiceTimes() == 0) {
+      newAddress = coordinatorDiscoverer_->updateAddress();
+    }
+    if (newAddress != serviceAddress_) {
+      LOG(INFO) << "Service Inventory changed to " << newAddress.getAddressStr()
+                << ":" << newAddress.getPort();
+      std::swap(serviceAddress_, newAddress);
+      client_ = std::make_shared<http::HttpClient>(
+          eventBaseThread_.getEventBase(),
+          serviceAddress_,
+          std::chrono::milliseconds(10'000),
+          pool_,
+          clientCertAndKeyPath_,
+          ciphers_);
+    }
+  } catch (const std::exception& ex) {
+    LOG(WARNING) << "Error occurred during updating service address: "
+                 << ex.what();
+    scheduleNext();
+    return;
+  }
+
+  auto [request, body] = httpRequest();
+
+  client_->sendRequest(request, body)
+      .via(eventBaseThread_.getEventBase())
+      .thenValue([this](auto response) {
+        auto message = response->headers();
+        if (message->getStatusCode() != http::kHttpAccepted ||
+            message->getStatusCode() != http::kHttpNoContent) {
```

Comment:
@pranjalssh This logic isn't correct. If announcement is OK, there is no content. We should not retry this case. https://github.com/prestodb/presto/pull/20754

Reply:
What does kHttpNoContent mean? Thanks!